src/software/ai/hl/stp/action/intercept_ball_action.cpp

```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
+    bool robot_in_intercept_position = false;
+    do
+    {
+        yield(std::make_unique<MoveIntent>(robot->id(), intercept_position,
+                                           (-ball.velocity()).orientation(), 0, 0,
+                                           DribblerEnable::ON, MoveType::NORMAL,
+                                           AutokickType::NONE, BallCollisionType::AVOID));
+
+        if (ball.velocity().length() < BALL_MOVING_SLOW_SPEED_THRESHOLD / 2.0)
         {
-            LOG(DEBUG) << "ball leaving field" << std::endl;
-            yield(std::make_unique<MoveIntent>(
-                robot->id(), point_ball_leaves_field.value(),
-                (ball.position() - robot->position()).orientation(), 0, 0,
-                DribblerEnable::ON, MoveType::NORMAL, AutokickType::NONE,
-                BallCollisionType::ALLOW));
+            restart();
         }
-        else
+
+        bool intercept_point_in_front_of_ball =
+            acuteAngle(ball.velocity(), intercept_position - ball.position()) <
+            Angle::quarter();
+        bool robot_at_intercept_point =
+            distance(robot->position(), intercept_position) < 0.015;
+        bool robot_stopped = robot->velocity().length() < 0.02;
+        robot_in_intercept_position =
+            intercept_point_in_front_of_ball && robot_at_intercept_point && robot_stopped;
+    } while (!robot_in_intercept_position);
+
+    // Wait for the ball to arrive at the robot
+    bool intercept_done = false;
+    do
+    {
+        yield(std::make_unique<MoveIntent>(robot->id(), intercept_position,
+                                           (-ball.velocity()).orientation(), 0, 0,
+                                           DribblerEnable::ON, MoveType::NORMAL,
+                                           AutokickType::NONE, BallCollisionType::ALLOW));
+
+        if (!intercept_done &&
+            ball.velocity().length() < BALL_MOVING_SLOW_SPEED_THRESHOLD / 2.0)
         {
-            // This is a fallback case that ideally should never be reached. We will only
-            // enter this case if the robot is not in front of the ball, and the ball is
-            // not within the field.
-            LOG(DEBUG) << "Moving to ball backup case" << std::endl;
-            yield(std::make_unique<MoveIntent>(
-                robot->id(), ball.position(),
-                (ball.position() - robot->position()).orientation(), 0, 0,
-                DribblerEnable::ON, MoveType::NORMAL, AutokickType::NONE,
-                BallCollisionType::ALLOW));
+            restart();
         }
-    } while (!robotHasPossession(ball.getPreviousStates(), robot->getPreviousStates()));
+        bool ball_very_close_to_robot_dribbler =
+            distance(ball.position(),
+                     robot->position() + Vector::createFromAngle(robot->orientation())
+                                             .normalize(DIST_TO_FRONT_OF_ROBOT_METERS)) <
+            0.2 + BALL_MAX_RADIUS_METERS;
```

Comment:
yeah it should be dynamic param or constat. This is the second time I've seen it

Reply:
made constant